### DISCUSSION TITLE
[dagster-pipes] Report / handle exceptions in the external process

### DISCUSSION THREAD

**Discussion by derek1032** [dagster-io/dagster#17986] [Q&A]
Hi Dagster team, 

Thank you for creating this amazing feature `pipes`. We've been testing the `pipes` functionality using `pipes_k8s_client`, however, we encounter some issues that may require your support to resolve:
1.  We discovered that when the external process raise exceptions, the behavior of the orchetration process is somewhat unpredictable. Sometimes it can successfuly capture external messages, but other times it displays a warning message and yields `STEP_FAILURE` event. The following is the warning message we received:
```
<img width="636" alt="Screen Shot 2023-11-14 at 12 17 25 PM" src="https://github.com/dagster-io/dagster/assets/13246451/5451aeb2-6f66-4aca-be9a-0ae35e528723">

[pipes] did not receive any messages from external process. Check stdout / stderr logs from the external process if possible.
PipesEnvContextInjector: Attempted to inject context directly, typically as an environment variable.
PipesK8sPodLogsMessageReader: Attempted to read messages by extracting them from kubernetes pod logs directly.
```
2. We've noticed there is an experimental change in dagster 1.5.7 about how pipes handles external exceptions. Thank you for developing this useful feature. After upgrading to 1.5.7, we can successfully see the message `external process pipes closed with exception`; However, the run can still end with  a `SUCCESS` event. We are wondering what would be the best practice to report the red engine event, which contains external error messages, if we are using `slack_on_run_failure_sensor`.
<img width="636" alt="Screen Shot 2023-11-14 at 12 17 25 PM" src="https://github.com/dagster-io/dagster/assets/13246451/94d83679-8d05-49da-8471-cc4801e89025">

3. If we are using `pipes_k8s_client` in partitioned ops instead of partitioned assets, the run will always produce a `STEP_FAILURE` event with error messages shown below. We would like to ask if current pipes feature is unable to be used on `ops` or if there is a way to resolve this issue?
<img width="1315" alt="Screen Shot 2023-11-14 at 12 43 51 PM" src="https://github.com/dagster-io/dagster/assets/13246451/aa162139-8a66-48d8-84a5-1c799d3ee964">

More info:
Dagster version: 1.5.7
dagster-pipes: 1.5.7

Thank you, 
Derek

**Comment by alangenfeld** [comment]
Thanks for the report, I will look in to this. 

> We discovered that when the external process raise exceptions, the behavior of the orchetration process is somewhat unpredictable

Just to confirm, when the external processes raises an exception does that bubble up and cause the container process to return a non zero exit code?

**Reply by alangenfeld** [reply]
I was unable to reproduce the pod not reporting as failed when an exception bubbles to the top

**Reply by derek1032** [reply]
@alangenfeld The images below are some screenshots related to the dagster run reported as successful when an exception occurring in the external process. The external process initiated by `pipes_k8s_client.run()` was terminated with a non zero exit code 1. However, in Dagster's web UI, we can see the dagster run was reported as successful. In addition, the dagster run container stopped with an exit code 0. If you need me to provide more information, please feel free to let me know. Thank you!
<img width="1139" alt="Screen Shot 2023-11-15 at 1 01 07 PM" src="https://github.com/dagster-io/dagster/assets/13246451/e692a9ba-cca7-43a8-b6ab-b669522aab11">
<img width="898" alt="Screen Shot 2023-11-15 at 1 01 59 PM" src="https://github.com/dagster-io/dagster/assets/13246451/3565682b-ad9e-4229-a2da-56bfeec46dca">
<img width="1139" alt="Screen Shot 2023-11-15 at 1 04 06 PM" src="https://github.com/dagster-io/dagster/assets/13246451/87db39c0-e5d2-4482-a4fa-667185265d61">
<img width="560" alt="Screen Shot 2023-11-15 at 5 54 59 PM" src="https://github.com/dagster-io/dagster/assets/13246451/901b3ff8-30d2-4e1a-9ca4-1bd795d9322d">

**Reply by alangenfeld** [reply]
hmm strange. This is the code I expect to raise (and what I see raise locally) https://github.com/dagster-io/dagster/blame/1.5.8/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L630-L635.

Can you share the body of `k8s_pipes_asset` ?

**Reply by derek1032** [reply]
Sure, here is the body of `k8s_pipes_asset`
<img width="975" alt="Screen Shot 2023-11-16 at 4 17 43 PM" src="https://github.com/dagster-io/dagster/assets/13246451/bf706f4b-f6b4-469c-8df6-9833721075f3">

**Reply by derek1032** [reply]
Hi @alangenfeld, I am suspecting when setting `wait_for_state=WaitForPodState.Ready`, the execution will break out of the `while loop` if the container status is `ready` and it may not reach the code you expect to raise `DagsterK8sError`
https://github.com/dagster-io/dagster/blob/41bc7864c350a3525decf5bb7630b24408a94a6f/python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py#L247-L259
https://github.com/dagster-io/dagster/blob/1f15855d6d137b4adcf586f08b012c262a062e61/python_modules/libraries/dagster-k8s/dagster_k8s/client.py#L579-L588

**Reply by alangenfeld** [reply]
ah yep good catch, will get a fix out

**Reply by alangenfeld** [reply]
https://github.com/dagster-io/dagster/pull/18076

**Comment by alangenfeld** [comment]
>  in partitioned ops instead of partitioned assets, the run will always produce a STEP_FAILURE event with error messages shown below. We would like to ask if current pipes feature is unable to be used on ops or if there is a way to resolve this issue?

`get_materialize_result()` is designed to work in the context of an `@asset`, in an `@op` you should be using `get_results()`. https://github.com/dagster-io/dagster/pull/17921 is exploring some improvements here, but we may do some further improvements to the error message.

**Reply by derek1032** [reply]
@alangenfeld Thank you for your information. I've switched to using `return pipes_k8s_client.run(...).get_results()` in the context of `@op`.  However, it appears that current behavior of the pipes will consistently yield a `STEP_FAILURE`with the following error messages. Unsure if the improvements you are exploring can address this issue and enable the use of `pipes_k8s_client` in `@ops`. <img width="1240" alt="Screen Shot 2023-11-15 at 5 20 58 PM" src="https://github.com/dagster-io/dagster/assets/13246451/b17ee1b1-0f96-4985-8b25-05a52ad8fb82">

**Reply by alangenfeld** [reply]
How are you defining your `@op` ? Will think on how to improve the ergonomics here but I expect this to work:
```
@op
def sample(context):
  yield from pipes_k8s_client.run().get_results()
  yield Output(...) 
```
and possibly this (not verified), assuming your op doesn't return anything
```
@op(out=Out(Nothing))
def sample(context):
  yield from pipes_k8s_client.run().get_results()
```

**Reply by derek1032** [reply]
The first method works, but the second one doesn't. Thank you!

**Reply by alangenfeld** [reply]
https://github.com/dagster-io/dagster/pull/18077
